optics-core/src/Optics/Internal/Optic.hs

```diff
@@ -210,25 +212,61 @@ class LabelOptic (name :: Symbol) k s t a b | name s -> k a
   -- corresponds to @'labelOptic' \@"foo"@.
   labelOptic :: Optic k NoIx s t a b
 
+-- | Type synonym for a type-preserving optic as overloaded label.
+type LabelOptic' name k s a = LabelOptic name k s s a a
+
+-- | Implements fallback behaviour in case there is no explicit 'LabelOptic'
+-- instance. This has a catch-all incoherent instance that merely yields an
+-- error message.  However, a downstream module can give a more specific
+-- instance that uses 'Generic' to construct an optic automatically.
+--
+-- To support this, the last parameter will be instantiated to 'True' if at
+-- least one of @s@ or @t@ has a 'Generic' instance.
+class GeneralLabelOptic (name :: Symbol) k s t a b (has_rep :: Bool) where
+  -- | Used to interpret overloaded label syntax in the absence of an explicit
+  -- 'LabelOptic' instance.
+  generalLabelOptic :: Optic k NoIx s t a b
+
+-- | If no instance matches, fall back on 'GeneralLabelOptic'.
+instance {-# OVERLAPPABLE #-}
+  (LabelOptic name k s t a b,  -- Needed to satisfy functional dependecy
+  GeneralLabelOptic name k s t a b (HasRep (Rep s) (Rep t))
+  ) => LabelOptic name k s t a b where
+  labelOptic = generalLabelOptic @name @k @s @t @a @b @(HasRep (Rep s) (Rep t))
+
+data Void0
+-- | If for an overloaded label @#label@ there is no instance starting with
+-- @LabelOptic "label"@, using it in the context of optics makes GHC immediately
+-- pick the generic overlappable instance (since no other instance could match),
+-- which ultimately leads to picking the instance with custom type error defined
+-- below. Prevent that (if only to be able to inspect most polymorphic types of
+-- @#foo % #bar@ or @view #foo@ in GHCi), by defining a dummy instance that
+-- matches all names, thus postponing the reduction.
+instance LabelOptic name An_Iso Void0 Void0 Void0 Void0 where
+  labelOptic = Optic id
+
+-- | This type family should be called with applications of 'Rep' on both sides,
+-- and will reduce to 'True' if at least one of them is defined; otherwise it is
+-- stuck.
+type family HasRep (s :: Type -> Type) (t :: Type -> Type) :: Bool
```

Comment:
Perhaps it would be better for it to be called `RepDefined` and return `()` instead of Bool? considering that `False` is never used.

Reply:
This would work. I feel somewhat uncomfortable returning an informative `()` as morally we ought to have an eta-law for it (and perhaps GHC one day will, although if it does that would probably break a lot of the type family tricks). `RepDefined` is a better name than `HasRep` in any case.